solr/solrj/src/java/org/apache/solr/client/solrj/cloud/autoscaling/PolicyHelper.java

```diff
@@ -429,87 +463,149 @@ private void release(SessionWrapper sessionWrapper) {
+        }
+      } else {
+        log.warn("returning unknown session {} ", sessionWrapper.getCreateTime());
+      }
     }
 
-
-    public SessionWrapper get(SolrCloudManager cloudManager) throws IOException, InterruptedException {
+    /**
+     * <p>Method returning an available session that can be used for {@link Status#COMPUTING}, either from the
+     * {@link #sessionWrapperSet} cache or by creating a new one. The status of the returned session is set to {@link Status#COMPUTING}.</p>
+     *
+     * Some waiting is done in two cases:
+     * <ul>
+     *   <li>A candidate session is present in {@link #sessionWrapperSet} but is still {@link Status#COMPUTING}, a random wait
+     *   is observed to see if the session gets freed to save a session creation and allow session reuse,</li>
+     *   <li>It is necessary to create a new session but there are already sessions in the process of being created, a
+     *   random wait is observed (if no waiting already occurred waiting for a session to become free) before creation
+     *   takes place, just in case one of the created sessions got used then {@link #returnSession(SessionWrapper)} in the meantime.</li>
+     * </ul>
+     *
+     * The random wait prevents the "thundering herd" effect when all threads needing a session at the same time create a new
+     * one even though some differentiated waits could have led to better reuse and less session creations.
+     *
+     * @param allowWait usually <code>true</code> except in tests that know there's no point in waiting because nothing
+     *                  will happen...
+     */
+    public SessionWrapper get(SolrCloudManager cloudManager, boolean allowWait) throws IOException, InterruptedException {
       TimeSource timeSource = cloudManager.getTimeSource();
+      long oldestUpdateTimeNs = TimeUnit.SECONDS.convert(timeSource.getTimeNs(), TimeUnit.NANOSECONDS) - SESSION_EXPIRY;
+      int zkVersion = cloudManager.getDistribStateManager().getAutoScalingConfig().getZkVersion();
+
       synchronized (lockObj) {
-        if (sessionWrapper.status == Status.NULL ||
-            sessionWrapper.zkVersion != cloudManager.getDistribStateManager().getAutoScalingConfig().getZkVersion() ||
-            TimeUnit.SECONDS.convert(timeSource.getTimeNs() - sessionWrapper.lastUpdateTime, TimeUnit.NANOSECONDS) > SESSION_EXPIRY) {
-          //no session available or the session is expired
-          return createSession(cloudManager);
-        } else {
+        SessionWrapper sw = getAvailableSession(zkVersion, oldestUpdateTimeNs);
+
+        // Best case scenario: an available session
+        if (sw != null) {
+          if (log.isDebugEnabled()) {
+            log.debug("reusing session {}", sw.getCreateTime());
+          }
+          return sw;
+        }
+
+        // Wait for a while before deciding what to do if waiting could help...
+        if ((creationsInProgress != 0 || hasCandidateSession(zkVersion, oldestUpdateTimeNs)) && allowWait) {
```

Comment:
Minor: put allowWait at beginning of check to short-circuit if allowWait is false

Reply:
Put it in purpose at the end: it's only true in tests